crates/sparrow-runtime/src/prepare/prepare_input_stream.rs

```diff
@@ -0,0 +1,280 @@
+    mut reader: BoxStream<'a, Result<RecordBatch, ArrowError>>,
+    config: &TableConfig,
+    raw_metadata: RawMetadata,
+    prepare_hash: u64,
+    slice: &Option<slice_plan::Slice>,
+) -> anyhow::Result<BoxStream<'a, error_stack::Result<(RecordBatch, RecordBatch), Error>>> {
+    // This is a "hacky" way of adding the 3 key columns. We may just want
+    // to manually do that (as part of deprecating `TableSchema`)?
+    let prepared_schema = TableSchema::try_from_data_schema(raw_metadata.table_schema.clone())?;
+    let prepared_schema = prepared_schema.schema_ref().clone();
+
+    // Add column behaviors for each of the 3 key columns.
+    let mut columns = Vec::with_capacity(prepared_schema.fields().len());
+    columns.push(ColumnBehavior::try_new_cast(
+        &raw_metadata.raw_schema,
+        &config.time_column_name,
+        &TimestampNanosecondType::DATA_TYPE,
+        false,
+    )?);
+    if let Some(subsort_column_name) = &config.subsort_column_name {
+        columns.push(ColumnBehavior::try_new_subsort(
+            &raw_metadata.raw_schema,
+            subsort_column_name,
+        )?);
+    } else {
+        columns.push(ColumnBehavior::try_default_subsort(prepare_hash)?);
+    }
+
+    columns.push(ColumnBehavior::try_new_entity_key(
+        &raw_metadata.raw_schema,
+        &config.group_column_name,
+        false,
+    )?);
+
+    // Add column behaviors for each column.  This means we include the key columns
+    // redundantly, but cleaning that up is a big refactor.
+    // See https://github.com/riptano/kaskada/issues/90
+    for field in raw_metadata.table_schema.fields() {
+        columns.push(ColumnBehavior::try_cast_or_reference_or_null(
+            &raw_metadata.raw_schema,
+            field,
+        )?);
+    }
+
+    // we've already checked that the group column exists so we can just unwrap it here
+    let (source_index, field) = raw_metadata
+        .raw_schema
+        .column_with_name(&config.group_column_name)
+        .unwrap();
+    let slice_preparer =
+        SlicePreparer::try_new(source_index, field.data_type().clone(), slice.as_ref())?;
+
+    let (_, entity_key_column) = raw_metadata
+        .raw_schema
+        .column_with_name(&config.group_column_name)
+        .with_context(|| "")?;
+
+    let mut metadata = PrepareMetadata::new(entity_key_column.data_type().clone());
+
+    Ok(async_stream::try_stream! {
+        while let Some(Ok(batch)) = reader.next().await {
+            // 1. Slicing may reduce the number of entities to operate and sort on.
+            let read_batch = slice_preparer.slice_batch(batch)?;
+            // 2. Prepare each of the columns by getting the column behavior result
+            let mut prepared_columns = Vec::new();
+            for c in columns.iter_mut() {
+                let result = c
+                    .get_result(&read_batch)
+                    .await?;
+
+                update_key_metadata(c, &read_batch, &result, &mut metadata)?;
```

Comment:
Same comment here... we shouldn't update key metadata for all columns... only the key hash column.

Reply:
same